AutomatedTesting/cmake/EngineFinder.cmake

```
@@ -7,35 +7,72 @@
 #
 #
 # {END_LICENSE}
-# This file is copied during engine registration. Edits to this file will be lost next
-# time a registration happens.
+# Edits to this file may be lost in upgrades. Instead of changing this file, use 
+# the 'engine_finder_cmake_path' key in your project.json or user/project.json to specify 
+# an alternate .cmake file to use instead of this one.
 
 include_guard()
 
-# Read the engine name from the project_json file
+include(cmake/VersionUtils.cmake)
+
+set(ENGINE_FINDER_VERBOSE FALSE CACHE BOOL "Increase EngineFinder.cmake verbosity.  Useful for debugging engine selection based on compatibility.")
+
+# Read the engine name from the shared or user project.json
 file(READ ${CMAKE_CURRENT_SOURCE_DIR}/project.json project_json)
 set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/project.json)
 
-string(JSON LY_ENGINE_NAME_TO_USE ERROR_VARIABLE json_error GET ${project_json} engine)
+string(JSON O3DE_ENGINE_NAME_TO_USE ERROR_VARIABLE json_error GET ${project_json} engine)
 if(json_error)
     message(FATAL_ERROR "Unable to read key 'engine' from 'project.json'\nError: ${json_error}")
 endif()
+ 
+# Let the user override the engine to use
+if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/user/project.json)
+    file(READ ${CMAKE_CURRENT_SOURCE_DIR}/user/project.json user_project_json)
+    set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/user/project.json)
+    string(JSON user_project_engine_name ERROR_VARIABLE json_error GET ${user_project_json} engine)
+    if(NOT json_error)
+        set(O3DE_ENGINE_NAME_TO_USE ${user_project_engine_name})
+    endif()
+    string(JSON user_engine_path ERROR_VARIABLE json_error GET ${user_project_json} "engine_path")
+endif()
+
+# Extract optional version specifier in the engine name 
+o3de_get_version_specifier_parts("${O3DE_ENGINE_NAME_TO_USE}" O3DE_ENGINE_NAME_TO_USE O3DE_ENGINE_SPECIFIER_OP O3DE_ENGINE_SPECIFIER_VERSION)
+
+# for backward compatibility
+set(LY_ENGINE_NAME_TO_USE ${O3DE_ENGINE_NAME_TO_USE})
 
 if(CMAKE_MODULE_PATH)
     foreach(module_path ${CMAKE_MODULE_PATH})
         if(EXISTS ${module_path}/Findo3de.cmake)
-            file(READ ${module_path}/../engine.json engine_json)
-            string(JSON engine_name ERROR_VARIABLE json_error GET ${engine_json} engine_name)
-            if(json_error)
-                message(FATAL_ERROR "Unable to read key 'engine_name' from 'engine.json'\nError: ${json_error}")
-            endif()
-            if(LY_ENGINE_NAME_TO_USE STREQUAL engine_name)
+            o3de_engine_compatible(${module_path}/../ ${O3DE_ENGINE_NAME_TO_USE} ${O3DE_ENGINE_SPECIFIER_OP} ${O3DE_ENGINE_SPECIFIER_VERSION} is_compatible engine_version ${ENGINE_FINDER_VERBOSE})
```

The first `${module_path}/..` should be put in double quotes.

If the module_path has a space in it such as `C:\o3de\My Engine`, then it will get parsed as two arguments.

Also I advise using [cmake_path(GET ... PARENT_PATH)](https://cmake.org/cmake/help/latest/command/cmake_path.html?highlight=cmake_path#get) to retrieve the parent directory 
```suggestion
            cmake_path(GET module_path PARENT_PATH candidate_engine_path)
            o3de_engine_compatible("${candidate_engine_path}" ${O3DE_ENGINE_NAME_TO_USE} ${O3DE_ENGINE_SPECIFIER_OP} ${O3DE_ENGINE_SPECIFIER_VERSION} is_compatible engine_version ${ENGINE_FINDER_VERBOSE})
```